### ISSUE TITLE
Adding event handlers in data templates pointing to methods of parent classes leads to memory leaks in WinUI 3

### ISSUE THREAD

**Issue by lhak** [microsoft/microsoft-ui-xaml#6894]
### Describe the bug

Adding event handlers (or referencing resources) in data templates (e.g. of a treeview item) that point to methods (or resources) of parent classes causes memory leaks (i.e. in c# the treeview items are not collected anymore). The same code in a UWP app does not show this behavior.

### Steps to reproduce the bug

Consider the following code from the repro project:

```xaml
  <TreeView x:Name="treeView" Height="400" ItemsSource="{x:Bind items}">
      <TreeView.Resources>
        <MenuFlyout x:Key="TreeItemContextMenu">
          <MenuFlyoutItem Text="Test" Click="MenuFlyoutItem_Click"></MenuFlyoutItem>

        </MenuFlyout>

        <!-- Adding a context menu through a style works properly -->
        <!--<Style TargetType="TreeViewItem">
          <Setter Property="ContextFlyout" Value="{StaticResource TreeItemContextMenu}"/>
        </Style>-->
      </TreeView.Resources>
      <TreeView.ItemTemplate>
        <DataTemplate x:DataType="local:Item">
          <TextBlock Text="{x:Bind Name, Mode=OneWay}" DoubleTapped="TextBlock_DoubleTapped" ContextFlyout="{StaticResource TreeItemContextMenu}"/>
          <!--<TextBlock Text="{x:Bind Name, Mode=OneWay}"/>-->
        </DataTemplate>
      </TreeView.ItemTemplate>
    </TreeView>
```

The referenced event handler and resource of the TextBlock will lead to memory leaks in a WinUI 3 project, but not in a UWP project. Adding an event handler to methods of the local:Item class prevents this. In the same way, adding the context flyout through a style will also avoid the memory leak.

This can be demonstrated using the attached repro project where tabview items containing this treeview can be added and removed. In the UWP app, the tabview items get collected when they are removed from the tabview, however this does not happen in the WinUI 3 project if the event handlers and resources mentioned above are referenced.

[ItemLeakRepro.zip](https://github.com/microsoft/microsoft-ui-xaml/files/8360963/ItemLeakRepro.zip)

### Expected behavior

No memory leak and no regression compared to a UWP app.

### NuGet package version

WinUI 3 - Windows App SDK 1.0

### Windows app type

- [ ] UWP
- [X] Win32

### Device form factor

Desktop

### Windows version

Windows 11 (21H2): Build 22000

**Comment by ojhad** [comment]
@RealTommyKlein This looks like a bug in the Xaml Compiler or CSWinrt?

**Comment by KenionX** [comment]
Noticed the same issue, anyone knowing a workarround for now?

**Comment by RealTommyKlein** [comment]
Thanks for the detailed repros @lhak !  I've been looking at this, and AFAICT this is due to some garbage collection and CSWinRT strangeness and not an actual memory leak.  I modified the repros to add/remove 100 at a time to exacerbate the issue, and ran through some scenarios taking memory snapshots at several points within a single run of the app.  Here's what I did/when I took the memory snapshots:

## Methodology

**Snapshot 1**: Taken immediately after `MainPage`/`MainWindow` has fully loaded
**Snapshot 2**: Taken after adding 100 `CustomTabViewItem`s using 1 press of the `Add` button
**Snapshot 3**: Taken after removing 100 `CustomTabViewItem`s using 1 press of the `Remove` button
**Snapshot 4**: Taken after adding 1000 `CustomTabViewItem`s using 10 presses of the `Add` button
**Snapshot 5**: Taken after removing 1000 `CustomTabViewItem`s using 10 presses of the `Remove` button
**Snapshot 6**: Taken after pressing `Remove` 5 times to trigger GC 5 separate times (even though there were no longer any `TabViewItem`s present).  (_note: I did this is a sanity check that GC is running completely after noticing odd behavior explained later_)

## Snapshot results
### UWP Windows.UI.Xaml
![uwp_memory](https://user-images.githubusercontent.com/45502154/163010617-cca4952f-080d-474e-ab56-71a42885f908.png)

This more or less behaves as expected, although we'd expect a larger memory takeback in Snapshot 3.  Additionally, Snapshot 6 shows more memory being freed even though Snapshot 5 should've already free'd everything.

### Desktop Windows App SDK 1.0.1
![desktop_memory](https://user-images.githubusercontent.com/45502154/163010681-ac3a17cd-3a11-4fc8-9891-762635f2ab66.png)

This shows bizarre behavior during Snapshot 3, actually increasing our memory when we are removing objects! We do see memory being freed up between Snapshots 2 and 3 and Snapshots 4 and 5 as expected, but Snapshot 6 still strangely frees more memory even though it all should've been freed in Snapshot 6.

## Results/Investigation
- **Why is Snapshot 6 able to free additional memory when it all should've been free'd in Snapshot 5**?
  - For some reasaon, `GC.Collect()` and `GC.WaitForPendingFinalizers()` don't actually free all the available memory, and require multiple runs to actually garbage collect all unused memory.  I found this could take as many as 5 runs before it had no effect.  Additionally a tight loop of calling them multiple times wasn't sufficient, they had to be invoked over time using the `Remove` button to space them out (although I imagine putting it on a different thread + adding a small `Thread.Sleep()` between calls would also work).
- **Why does Snapshot 3 in the Windows App SDK actually increase memory used, when it should be reclaiming memory?**
  - The snapshot showed the increase was due to newly CSWinRT related objects.  I believe casting the `TabItems` contents to `CustomTabViewItem` calls into CSWinRT to fulfill the cast, and results in a bunch of new CSWinRT-related objects being created.  Because of the issue with garbage collection not actually collecting all the unused memory, these new objects aren't immediately cleared during Snapshot 3's garbage collection.

After looking into these, I did one more smaller test with two snapshots:
**Snapshot 1**: After adding and then removing 100 items, then clicking `Remove` 5 times to fully garbage collect everything.  This is to get the app "warm" so objects which may've been created during normal app use are initialized.
**Snapshot 2**: Adding and then removing 1000 items 3 times in a row, then clicking `Remove` 5 times to fully garbage collect everything.  This is to fully exercise the app + completely reclaim unused memory with garbage collection.

In Windows.UI.Xaml UWP, the heap did not contain anything new or have additional refcounts between the snapshots, as expected.  In Desktop Windows App SDK, the only new things on the heap were from `GCHeapHash`, which I believe is a [data structure used as part of heap management](https://github.com/Potapy4/dotnet-coreclr/blob/master/src/vm/gcheaphashtable.h) (so not surprising it grew after needing to store 1000 objects at a time).  Otherwise there were no new refcounts, and there were no other leaked objects, including `CustomTabViewItem`s:
![winui_heap](https://user-images.githubusercontent.com/45502154/163022150-92e126b5-95c9-45fa-9012-460eae68f138.png)

## Conclusion
Because of the strange garbage collection behavior applications' memory footprint may continue increasing unexpectedly when it shouldn't be.  But I don't believe there's a legitimate memory leak here, and after enough automatically (or manually) triggered garbage collections have happened the unused memory will be reclaimed by garbage collection.  If you are finding different, let me know 😄

**Comment by KenionX** [comment]
@RealTommyKlein why not checking if the objects are actually deallocated? You could check how many instances of your object exist. Or simply put a **deconstructor** on your object, it cleary shows that the objects are **not** deallocated and persist in memory even after multiple GC runs. This is by definition a memory leak

**Comment by RealTommyKlein** [comment]
@KenionX - they are deallocated in the repro posted in the original issue.  The repro has a `CustomTabViewItem` type which increments/decrements a static instance counter in its constructor/destructor.  All of the `CustomTabViewItem` instances are destroyed within a few GC iterations.  If you have a different repro which does have a leak which persists, I can take a look.

**Comment by KenionX** [comment]
@RealTommyKlein 
I created a project to reproduce it, adding the "DoubleTapped" event causes the memory leak, after 15gb allocation I stopped the test, if you remove the event it will deconstruct all objects without issues.

https://github.com/KenionX/Winui3MemoryLeak

I noticed that the issue will only occure when the control was visualized once, if you allocated 1000 objects with a loop it will deallocate some of them.

**Comment by RealTommyKlein** [comment]
@KenionX - thanks, can repro with yours!  Will take a look.

**Comment by RealTommyKlein** [comment]
This has been fixed for the upcoming Windows App SDK 1.2 Preview 3 release (verified with both @KenionX 's repro and @lhak 's unmodified repro).  It has the same root cause as #6123, where `DataTemplate`s had an issue when instantiating their content which would leak the root control if an event subscription in the template used a method defined on the root control.
